### PR TITLE
:honeybee: (svg tester) test hidden chart views

### PR DESCRIPTION
**Pull request by sophiamersmann**
I love our SVG tester!! It has saved me many times, and I wouldn't want to live without it ❤️ 

### Summary

- The SVG tester currently tests the default chart view for each published chart
- This PR extends the SVG tester:
    - It adds the ability to test a particular chart view in an ad-hoc way
    - It adds systematic testing for all possible chart views of a subset of charts

### Motivation

- Let's say I work on a very specific chart view, something like StackedArea charts that are faceted by metric and are in relative mode
- At the moment, I can't trust that any of the published charts in their default view are actually configured this way
- As a result, the SVG tester's green checkmark is useful in the sense that I apparently haven't broken anything unrelated, but it also doesn't really give me useful information for the task that I'm working on
- What I would want from an automatic testing tool is to answer two questions:
    1. Is there any faceted StackedArea chart in relative mode that I just broke? (i.e. verify newly introduced behaviour is correct)
    2. Have I broken anything else that is (seemingly) unrelated? (i.e. verify old behaviour is preserved)
- Our SVG tester partly answers the second question, but can't answer the first:
    1. The first question can't be answered since our SVG tester doesn't have visibility on chart views that are hidden behind interactions
    2. For the second question, the current SVG tester gives a good indication – although it likely doesn't cover _all_ possible configurations for each chart type 

### Details

We can't test all chart views of all published charts. Instead, I added
- automatic testing of all possible chart views for a (small) subset of charts via a new GitHub action
- a script for local/ad-hoc testing that checks a specific chart view for all (or any subset) of charts

#### Local testing of a specific chart view

- The export and verify script both accept a new option, `--query-str`/`-q`
    - The given query string is passed to Grapher, and the exported SVG captures that particular view
    - Example: `--query-str facet=metric&stackMode=relative`
- Since we don't have reference SVGs for every chart view and for all charts, we need to create reference SVG first and then verify them in a second step
- `verify-against-master.sh` does exactly that. It's a utility script to:
    - Stash current changes
    - Checkout master
    - Export SVGs on master
    - Checkout last branch
    - Verify SVGs on the current branch
    - Create an HTML report if there are differences
- Usually, it doesn't make sense to verify all charts in this way (generating all SVGs twice would take some time, and most query string parameter combinations only make sense for a subset of charts)
- That's why the export and verify scripts both accept a range of options to choose a set of charts to process, including:
    - `--ids`/`-c`: List of IDs or ID ranges
    - `--chart-types`/`-t`: List of chart types

For example, I could check if any faceted StackedArea charts in relative mode have changed by running:
```bash
verify-against-master.sh -t=StackedArea -q=facet=metric&stackMode=relative
```

If running for the second time, the reference SVGs don't need to be re-generated. I can skip the export check like this:
```bash
verify-against-master.sh -t=StackedArea -q=facet=metric&stackMode=relative --skip-export
```

Relying on the query string to force Grapher into a particular state has a couple of gotchas:
- Not all generated SVGs are necessarily live since we don't check whether a view is reachable for a user (e.g. via the query string, you can dictate a chart to be in relative mode, although the settings UI doesn't actually include the relative mode toggle)
- Some generated SVGs might show "fallback" charts (e.g. faceting a line chart with a single dimension displays an un-faceted chart)

The first point is something to keep in mind. The second doesn't have to be a bad thing. Let's say a particular view is forbidden for a chart type. In that case, I would expect to see a fallback. If that unintentionally changes, then that's useful information.

#### New GitHub action: Testing all possible chart views for a subset of charts

- The export and verify script both accept a new option, `--all-views`
    - If specified, the script generates SVGs for all possible chart views for a given chart

How all possible chart views are generated:
- For each chart type, all possible configurations are listed in `chart-configurations.ts` and translated into a list of query strings for each chart type
    - Since we rely on query strings, this approach suffers from the same gotchas described above
    - Alternatively, I could have – for each given chart – figured out which configurations are valid and live and then only generated those, but that would have been much harder, and I realised that most charts of the same type have similar views, so it didn't seem worth it

Which subset of charts is tested:
- For each chart type, I chose to test the top 30 most viewed charts (~220 charts and ~3300 chart views)
    - It's important to have a diverse set of charts. For example, we need chart configurations with multiple dimensions to test faceting properly. The most viewed charts turn out to be pretty diverse, so I stuck with this simple approach.

How the GitHub action works:
- Basically, exactly like the current SVG tester workflow works
- The chart IDs to test are dumped into a file named `most-viewed-charts.txt` in the `owid-grapher-svgs` repo by the `dump-chart-ids.ts` script
- The reference SVGs are written to the `all-views/svg` folder in the `owid-grapher-svgs` repo
- The GitHub action itself is added in the follow-up PR, #3062 (but the necessary code is added in this one)

### Other changes

- I added a `--random` flag to both the export and verify script. If specified, a random subset of charts is tested
- I added lazy loading of images to the `create-compare-view.ts` script
    - I deleted the `diff-generator.ts` script. It did basically what the `create-compare-view.ts` script does but offered to generate one HTML file per chart, presumably for cases where we have many differences and one big HTML file takes forever to load. I hope that adding lazy loading to the `create-compare-view` script is good enough to deal with such cases
- I deleted the `SVGTester.tsx` file. It doesn't seem to be in use. If there is a reason for keeping it, I'm happy to bring it back!

### CODE REVIEW

**Comment by sophiamersmann** [comment]
Current dependencies on/for this PR:
* `master`
  * **PR #3043** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3043?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/owid/owid-grapher/3043?utm_source=stack-comment).

**Comment by sophiamersmann** [comment]
@danyx23 the `verify-against-master.sh` script is a little difficult to test because it checkouts the master, and then runs the old version of the `export` script on master, which doesn't have all params that the new version has (e.g. the `--query-str` param)

**Comment by sophiamersmann** [comment]
@danyx23 the `verify-against-master.sh` script is a little difficult to test because it checkouts the master, and then runs the old version of the `export` script on master, which doesn't have all params that the new version has (e.g. the `--query-str` param)

**Review by danyx23 — COMMENTED** [review]
Really nice work! I found a few minor things that I think would be nice to improve but overall this is already great and super useful. Thank a lot for giving this some love!

File: db/model/Chart.ts [hunk 1]
```diff
@@ -328,3 +328,22 @@ export const getGrapherById = async (grapherId: number): Promise<any> => {
...
+    count = 10
+): Promise<number[]> => {
+    const ids = await db.queryMysql(
+        `SELECT c.id
+        FROM analytics_pageviews a
+        JOIN charts c ON c.slug = SUBSTRING_INDEX(a.url, '/', -1)
+        WHERE a.url LIKE "https://ourworldindata.org/grapher/%"
+            AND c.type = ?
+            AND c.config ->> "$.isPublished" = "true"
+            and (c.config ->> "$.hasChartTab" = "true" or c.config ->> "$.hasChartTab" is null)
```
[the remark this answers is not in the document]
Comment (sophiamersmann) [reply db/model/Chart.ts:343]: yes – what is currently tested are all _chart_ views, ignoring maps

File: devTools/svgTester/chart-configurations.ts [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+) as Record<ChartTypeName, string[]>
+
+// adapted from https://stackoverflow.com/questions/12303989/cartesian-product-of-multiple-arrays-in-javascript
+function cartesian(matrix: any[][]) {
+    if (matrix.length === 0) return []
+    if (matrix.length === 1) return matrix[0].map((i) => [i])
+    return matrix.reduce((acc, curr) =>
+        acc.flatMap((i: any) => curr.map((j: any) => [i, j].flat()))
+    )
+}
```
[the remark this answers is not in the document]
Comment (sophiamersmann) [reply devTools/svgTester/chart-configurations.ts:157]: that sounds reasonable!

(I think technically, not all list items have to be of the same type, e.g. `[[1, 2], ["a", null]]` would be valid input, but let's type it stricter anyway)

File: devTools/svgTester/verify-against-master.sh [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        # store original state
+        stashChanges
+
+        # run export script on master
+        checkoutMaster
+        runExportScript
+
+        # restore original state
+        checkoutLastBranch
+        unstashChanges
```
[the remark this answers is not in the document]
Comment (sophiamersmann) [reply devTools/svgTester/verify-against-master.sh:152]: oh, good catch! that was unintentional

File: devTools/svgTester/verify-against-master.sh [hunk 4]
```diff
@@ -0,0 +1,160 @@
+#!/usr/bin/env bash
+
+usage() {
+  echo -e "Usage: ./$(basename $0) [-c | --ids] [-t | --chart-types] [-q | --query-str] [--all-views] [--skip-export] [-h | --help]
+
+Export Grapher charts from master and verify them against the current branch.
+
+Options:
+    --ids, -c           Config IDs passed to export-graphs.js and verify-graphs.js
```
[the remark this answers is not in the document]
Comment (sophiamersmann) [reply devTools/svgTester/verify-against-master.sh:9]: good point!

File: devTools/svgTester/verify-against-master.sh [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+        checkoutMaster
+        runExportScript
+
+        # restore original state
+        checkoutLastBranch
+        unstashChanges
+    fi
+    
+    # verify grapher charts and create an html report if there are differences
+    runVerifyScript || createHTMLReport
```
[the remark this answers is not in the document]
Comment (sophiamersmann) [reply devTools/svgTester/verify-against-master.sh:156]: Good point. I'll add the verbose flag to the verify script – then the script can speak for itself

**Pull request merged**